crossdock/server/server.py

```diff
@@ -130,79 +166,66 @@ def handle_trace_request(self, http_request, trace_request, span_handler, respon
-        url = "http://%s:%s/join_trace" % (downstream.host, downstream.port)
-        body = serializer.downstream_to_json(downstream.downstream)
-
-        req = tornado.httpclient.HTTPRequest(url=url, method="POST",
-                                             headers={"Content-Type": "application/json"},
-                                             body=body)
-        http_client.before_http_request(tornado_http.TornadoRequestWrapper(request=req),
-                                        lambda: span)
-        client = tornado.httpclient.AsyncHTTPClient()
-        client.fetch(req, handle_response)
-
-    def call_downstream_tchannel(self, span, trace_request, trace_response, response_writer):
-
-        tr = tchannelNotImplResponse()
-        response_writer.write(serializer.obj_to_json(tr))
-        response_writer.finish()
-
-    def real_call_downstream_tchannel(self, span, trace_request, trace_response, response_writer):
-
-        def handle_response(f):
-            response = f.result()
-            trace_response.downstream = response.body
-
-            response_writer.write(serializer.obj_to_json(trace_response))
-            response_writer.finish()
-
-        downstream = trace_request.downstream
-        # XXX cache these
-        service = thrift.load(idl_path, service=downstream.serviceName)
-
-        jtr = JoinTraceRequest(trace_request.serverRole, downstream.downstream)
-        jtr = join_trace_request_to_thriftrw(service, jtr)
-        # with context.RequestContext(span):
-        with context.request_context(span):
-            f = tchannel.thrift(service.TracedService.joinTrace(jtr),
-                                hostport="%s:%s" % (downstream.host, downstream.port))
-        tornado.ioloop.IOLoop.current().add_future(f, handle_response)
+                response_writer.finish()
+
+
+@tornado.gen.coroutine
+def call_downstream(span, downstream):
+    if downstream.transport == Transport.HTTP:
+        downstream_trace_resp = yield call_downstream_http(span, downstream)
+    elif downstream.transport == Transport.TCHANNEL:
+        downstream_trace_resp = yield call_downstream_tchannel(downstream)
+    else:
+        raise UnknownTransportException('%s' % downstream.transport)
+    raise tornado.gen.Return(downstream_trace_resp)
+
+
+@tornado.gen.coroutine
+def call_downstream_http(span, downstream):
+    url = 'http://%s:%s/join_trace' % (downstream.host, downstream.port)
+    body = serializer.join_trace_request_to_json(downstream.downstream, downstream.serverRole)
+
+    req = tornado.httpclient.HTTPRequest(url=url, method='POST',
+                                         headers={'Content-Type': 'application/json'},
+                                         body=body)
+    http_client.before_http_request(tornado_http.TornadoRequestWrapper(request=req),
+                                    lambda: span)
+    client = tornado.httpclient.AsyncHTTPClient()
+    future = yield client.fetch(req)
```

Comment:
something odd here, yield should not be returning future, but the actual data. Perhaps variable name is wrong, since you call `.body` on it - it's probably just an http response.